### PR TITLE
Fixed discovery tests

### PR DESCRIPTION
**Pull request by sghai**
Discovery tests have been failing since long w/ zstream jobs.. This PR should fix all of them. I tried to pull changes/enhancements from master and those are relevant to zstream. 

These  changes not only fix the existing tests but also allow us to expand the coverage.

### CODE REVIEW

**Comment by sghai** [comment]
Test results:
```console

[sghai@dhcp223-111 robottelo]$ py.test -v tests/foreman/ui/test_discoveredhost.py::DiscoveryTestCase
================================================================ test session starts =================================================================
platform linux2 -- Python 2.7.12, pytest-2.9.2, py-1.4.32, pluggy-0.3.1 -- /usr/bin/python
cachedir: .cache
rootdir: /home/sghai/QE/robottelo, inifile: 
plugins: services-1.1.14, cov-2.3.1
collected 42 items 
2017-04-24 18:50:10 - conftest - DEBUG - Found WONTFIX in decorated tests ['1110476', '1156555', '1269196', '1402826', '1245334', '1221971', '1217635', '1226425', '1199150', '1204686', '1267224', '1103157', '1230902', '1214312', '1079482']

2017-04-24 18:50:10 - conftest - DEBUG - Collected 42 test cases


tests/foreman/ui/test_discoveredhost.py::DiscoveryTestCase::test_custom_facts_discovery <- robottelo/decorators/__init__.py PASSED
tests/foreman/ui/test_discoveredhost.py::DiscoveryTestCase::test_negative_add_fact <- robottelo/decorators/__init__.py SKIPPED
tests/foreman/ui/test_discoveredhost.py::DiscoveryTestCase::test_negative_create_discovery_rule <- robottelo/decorators/__init__.py SKIPPED
tests/foreman/ui/test_discoveredhost.py::DiscoveryTestCase::test_negative_pxe_less_discovery_without_dhcp <- robottelo/decorators/__init__.py SKIPPED
tests/foreman/ui/test_discoveredhost.py::DiscoveryTestCase::test_negative_validate_network_config_screen <- robottelo/decorators/__init__.py SKIPPED
tests/foreman/ui/test_discoveredhost.py::DiscoveryTestCase::test_positive_add_fact_column <- robottelo/decorators/__init__.py PASSED
tests/foreman/ui/test_discoveredhost.py::DiscoveryTestCase::test_positive_auto_provision_all <- robottelo/decorators/__init__.py SKIPPED
tests/foreman/ui/test_discoveredhost.py::DiscoveryTestCase::test_positive_auto_provision_host_with_rule <- robottelo/decorators/__init__.py SKIPPED
tests/foreman/ui/test_discoveredhost.py::DiscoveryTestCase::test_positive_delete <- robottelo/decorators/__init__.py PASSED
tests/foreman/ui/test_discoveredhost.py::DiscoveryTestCase::test_positive_delete_from_facts <- robottelo/decorators/__init__.py PASSED
tests/foreman/ui/test_discoveredhost.py::DiscoveryTestCase::test_positive_delete_multiple <- robottelo/decorators/__init__.py PASSED
tests/foreman/ui/test_discoveredhost.py::DiscoveryTestCase::test_positive_discover_pxe_less_with_efi_host_interatively <- robottelo/decorators/__init__.py SKIPPED
tests/foreman/ui/test_discoveredhost.py::DiscoveryTestCase::test_positive_discover_pxe_less_with_efi_host_unattended <- robottelo/decorators/__init__.py SKIPPED
tests/foreman/ui/test_discoveredhost.py::DiscoveryTestCase::test_positive_discovery_manager_role <- robottelo/decorators/__init__.py SKIPPED
tests/foreman/ui/test_discoveredhost.py::DiscoveryTestCase::test_positive_discovery_reader_role <- robottelo/decorators/__init__.py SKIPPED
tests/foreman/ui/test_discoveredhost.py::DiscoveryTestCase::test_positive_multi_provision_with_rule_limit <- robottelo/decorators/__init__.py SKIPPED
tests/foreman/ui/test_discoveredhost.py::DiscoveryTestCase::test_positive_provision <- robottelo/decorators/__init__.py SKIPPED
tests/foreman/ui/test_discoveredhost.py::DiscoveryTestCase::test_positive_provision_from_facts <- robottelo/decorators/__init__.py SKIPPED
tests/foreman/ui/test_discoveredhost.py::DiscoveryTestCase::test_positive_provision_multi_host_with_rule <- robottelo/decorators/__init__.py SKIPPED
tests/foreman/ui/test_discoveredhost.py::DiscoveryTestCase::test_positive_provision_with_rule_priority <- robottelo/decorators/__init__.py SKIPPED
tests/foreman/ui/test_discoveredhost.py::DiscoveryTestCase::test_positive_provision_without_auto_provision <- robottelo/decorators/__init__.py SKIPPED
tests/foreman/ui/test_discoveredhost.py::DiscoveryTestCase::test_positive_pxe_based_discovery <- robottelo/decorators/__init__.py PASSED
tests/foreman/ui/test_discoveredhost.py::DiscoveryTestCase::test_positive_pxe_less_multi_nic_with_dhcp_interactively <- robottelo/decorators/__init__.py SKIPPED
tests/foreman/ui/test_discoveredhost.py::DiscoveryTestCase::test_positive_pxe_less_multi_nic_with_dhcp_unattended <- robottelo/decorators/__init__.py PASSED
tests/foreman/ui/test_discoveredhost.py::DiscoveryTestCase::test_positive_pxe_less_multi_nic_without_dhcp_interactively <- robottelo/decorators/__init__.py SKIPPED
tests/foreman/ui/test_discoveredhost.py::DiscoveryTestCase::test_positive_pxe_less_multi_nic_without_dhcp_unattended <- robottelo/decorators/__init__.py SKIPPED
tests/foreman/ui/test_discoveredhost.py::DiscoveryTestCase::test_positive_pxe_less_with_dhcp_interactively <- robottelo/decorators/__init__.py SKIPPED
tests/foreman/ui/test_discoveredhost.py::DiscoveryTestCase::test_positive_pxe_less_with_dhcp_semiauto <- robottelo/decorators/__init__.py SKIPPED
tests/foreman/ui/test_discoveredhost.py::DiscoveryTestCase::test_positive_pxe_less_with_dhcp_unattended <- robottelo/decorators/__init__.py PASSED
tests/foreman/ui/test_discoveredhost.py::DiscoveryTestCase::test_positive_pxe_less_without_dhcp_interactively <- robottelo/decorators/__init__.py SKIPPED
tests/foreman/ui/test_discoveredhost.py::DiscoveryTestCase::test_positive_pxe_less_without_dhcp_semiauto <- robottelo/decorators/__init__.py SKIPPED
tests/foreman/ui/test_discoveredhost.py::DiscoveryTestCase::test_positive_pxe_less_without_dhcp_unattended <- robottelo/decorators/__init__.py SKIPPED
tests/foreman/ui/test_discoveredhost.py::DiscoveryTestCase::test_positive_pxe_multi_nic_unattended <- robottelo/decorators/__init__.py PASSED
tests/foreman/ui/test_discoveredhost.py::DiscoveryTestCase::test_positive_reboot <- robottelo/decorators/__init__.py FAILED
tests/foreman/ui/test_discoveredhost.py::DiscoveryTestCase::test_positive_refresh_facts_pxe <- robottelo/decorators/__init__.py PASSED
tests/foreman/ui/test_discoveredhost.py::DiscoveryTestCase::test_positive_refresh_facts_pxe_less <- robottelo/decorators/__init__.py PASSED
tests/foreman/ui/test_discoveredhost.py::DiscoveryTestCase::test_positive_update_default_location <- robottelo/decorators/__init__.py PASSED
tests/foreman/ui/test_discoveredhost.py::DiscoveryTestCase::test_positive_update_default_org <- robottelo/decorators/__init__.py PASSED
tests/foreman/ui/test_discoveredhost.py::DiscoveryTestCase::test_positive_update_discovery_prefix <- robottelo/decorators/__init__.py SKIPPED
tests/foreman/ui/test_discoveredhost.py::DiscoveryTestCase::test_positive_update_discovery_rule <- robottelo/decorators/__init__.py SKIPPED
tests/foreman/ui/test_discoveredhost.py::DiscoveryTestCase::test_positive_update_name <- robottelo/decorators/__init__.py SKIPPED
tests/foreman/ui/test_discoveredhost.py::DiscoveryTestCase::test_positive_validate_pxe_less_discovery_status_screen <- robottelo/decorators/__init__.py SKIPPED

```

File: robottelo/api/utils.py [hunk 1]
```diff
@@ -2,10 +2,19 @@
...
 import time
 
+from fauxfactory import gen_string
 from inflector import Inflector
-from nailgun import entities
+from nailgun import entities, entity_mixins
 from robottelo import ssh
 from robottelo.constants import PERMISSIONS_WITH_BZ
+from robottelo.config import settings
+from robottelo.constants import (
```
Comment (oshtaier) [inline robottelo/api/utils.py]: please, put in one statement with `from robottelo.constants import PERMISSIONS_WITH_BZ`

File: robottelo/api/utils.py [hunk 2]
```diff
@@ -293,3 +302,242 @@ def create_role_permissions(role, permissions_types_names):  # pragma: no cover
...
+    # Create a new Life-Cycle environment
+    lc_env = entities.LifecycleEnvironment(organization=org).create()
+    # Create a Product, Repository for custom RHEL6 contents
+    product = entities.Product(organization=org).create()
+    repo = entities.Repository(
+        product=product,
+        url=settings.rhel7_os
+    ).create()
+
+    # Increased timeout value for repo sync
```
Comment (oshtaier) [inline robottelo/api/utils.py]: `# Increased timeout value for repo sync and cv publishing and promotion`

File: robottelo/api/utils.py [hunk 3]
```diff
@@ -293,3 +302,242 @@ def create_role_permissions(role, permissions_types_names):  # pragma: no cover
...
+        content_view = entities.ContentView(organization=org).create()
+        content_view.repository = [repo]
+        content_view = content_view.update(['repository'])
+        content_view.publish()
+        content_view = content_view.read()
+        promote(content_view.version[0], lc_env.id)
+    finally:
+        entity_mixins.TASK_TIMEOUT = old_task_timeout
+    # Search for puppet environment and associate location
+    environment = entities.Environment(
```
Comment (oshtaier) [inline robottelo/api/utils.py]: Do we have at least one puppet environment for all organizations by default?

File: robottelo/api/utils.py [hunk 4]
```diff
@@ -293,3 +302,242 @@ def create_role_permissions(role, permissions_types_names):  # pragma: no cover
...
+        query={
+            u'search': u'name={0}'.format(
+                settings.server.hostname)
+        }
+    )
+    proxy = proxy[0].read()
+    proxy.location.append(loc)
+    proxy = proxy.update(['location'])
+    proxy.organization.append(org)
+    proxy = proxy.update(['organization'])
```
Comment (oshtaier) [inline robottelo/api/utils.py]: I think we can update proxy by one statement like `proxy.update(['location', 'organization'])`

File: robottelo/api/utils.py [hunk 5]
```diff
@@ -293,3 +302,242 @@ def create_role_permissions(role, permissions_types_names):  # pragma: no cover
...
+            provider=u'libvirt',
+            url=resource_url,
+            set_console_password=False,
+            display_type=u'VNC',
+            location=[loc.id],
+            organization=[org.id],
+        ).create()
+
+    # Get the Partition table ID
+    ptable = entities.PartitionTable().search(
```
Comment (oshtaier) [inline robottelo/api/utils.py]: Shouldn't we assign ptable to created organization and location?

File: robottelo/api/utils.py [hunk 6]
```diff
@@ -293,3 +302,242 @@ def create_role_permissions(role, permissions_types_names):  # pragma: no cover
...
+    pxe_template.operatingsystem.append(os)
+    pxe_template.organization.append(org)
+    pxe_template.location.append(loc)
+    pxe_template = pxe_template.update(
+        ['location', 'operatingsystem', 'organization']
+    )
+
+    # Get the arch ID
+    arch = entities.Architecture().search(
+        query={u'search': u'name="x86_64"'}
```
Comment (oshtaier) [inline robottelo/api/utils.py]: we have `DEFAULT_ARCHITECTURE = 'x86_64'` in master. Maybe, you like to move that change to 6.2.z

File: robottelo/api/utils.py [hunk 7]
```diff
@@ -293,3 +302,242 @@ def create_role_permissions(role, permissions_types_names):  # pragma: no cover
...
+    )[0].read()
+
+    # Get the media and update its location
+    media = entities.Media(organization=[org]).search()[0].read()
```
Comment (oshtaier) [inline robottelo/api/utils.py]: so media should be populated automatically?

File: robottelo/ui/discoveryrules.py [hunk 8]
```diff
@@ -28,9 +30,22 @@ def _configure_discovery(self, hostname=None, host_limit=None,
...
+            self.configure_entity(
+                organizations,
+                FILTER['discovery_rule_org'],
+                tab_locator=tab_locators['tab_org'],
+            )
 
     def create(self, name, search_rule, hostgroup, hostname=None,
-               host_limit=None, priority=None, enabled=True):
+               host_limit=None, priority=None, locations=None,
+               organizations=None, select=True, enabled=True):
```
Comment (oshtaier) [inline robottelo/ui/discoveryrules.py]: shouldn't we also change
`self._configure_discovery(hostname, host_limit, priority, enabled, location, organization)`

**Comment by sghai** [comment]
@oshtaier thank you for all review comments.. I cherry-picked the changes from master but all changes were not relevant so some left outs were there. I updated the code as per comments. 

However, I'll re-run the tests once for both discoveredhosts and discovery rules to make sure everything is working.

**Comment by sghai** [comment]
test results:

```
[sghai@dhcp223-111 robottelo]$ py.test -v tests/foreman/ui/test_discoveryrule.py::DiscoveryRuleTestCase -k test_positive_create_with_hostname
================================================================ test session starts =================================================================
platform linux2 -- Python 2.7.12, pytest-2.9.2, py-1.4.32, pluggy-0.3.1 -- /usr/bin/python
cachedir: .cache
rootdir: /home/sghai/QE/robottelo, inifile: 
plugins: services-1.1.14, cov-2.3.1
collected 24 items 
2017-04-28 20:15:28 - conftest - DEBUG - Found WONTFIX in decorated tests ['1110476', '1156555', '1269196', '1402826', '1245334', '1221971', '1217635', '1226425', '1199150', '1204686', '1267224', '1103157', '1230902', '1214312', '1079482']

2017-04-28 20:15:28 - conftest - DEBUG - Collected 24 test cases


tests/foreman/ui/test_discoveryrule.py::DiscoveryRuleTestCase::test_positive_create_with_hostname <- robottelo/decorators/__init__.py PASSED

=========================================== 23 tests deselected by '-ktest_positive_create_with_hostname' ============================================
====================================================== 1 passed, 23 deselected in 65.48 seconds ======================================================
```

**Comment by sghai** [comment]
In above test results the failed test was `test_positive_reboot` and its fixed now.

```console

 robottelo]$ py.test -v tests/foreman/ui/test_discoveredhost.py::DiscoveryTestCase -k test_positive_reboot
================================================================ test session starts =================================================================
platform linux2 -- Python 2.7.12, pytest-2.9.2, py-1.4.32, pluggy-0.3.1 -- /usr/bin/python
cachedir: .cache
rootdir: /home/sghai/QE/robottelo, inifile: 
plugins: services-1.1.14, cov-2.3.1
collected 42 items 
2017-04-28 21:02:07 - conftest - DEBUG - Found WONTFIX in decorated tests ['1110476', '1156555', '1269196', '1402826', '1245334', '1221971', '1217635', '1226425', '1199150', '1204686', '1267224', '1103157', '1230902', '1214312', '1079482']

2017-04-28 21:02:07 - conftest - DEBUG - Collected 42 test cases


tests/foreman/ui/test_discoveredhost.py::DiscoveryTestCase::test_positive_reboot <- robottelo/decorators/__init__.py PASSED

================================================== 41 tests deselected by '-ktest_positive_reboot' ===================================================
===================================================== 1 passed, 41 deselected in 696.43 seconds ======================================================
```

**Comment by sghai** [comment]
Can someone please take another look ? This will definitely improve 6.2.z tier3 results.

**Comment by sghai** [comment]
@oshtaier Hey.. can we please merge this PR ?. So far all your comments have been addressed. Please take a look.

File: robottelo/config/settings.py [hunk 9]
```diff
@@ -742,6 +743,7 @@ def _read_robottelo_settings(self):
         self.project = self.reader.get('robottelo', 'project', 'sat')
         self.rhel6_repo = self.reader.get('robottelo', 'rhel6_repo', None)
         self.rhel7_repo = self.reader.get('robottelo', 'rhel7_repo', None)
+        self.rhel7_os = self.reader.get('robottelo', 'rhel7_os', None)
```
Comment (abalakh) [inline robottelo/config/settings.py]: Missing `rhel6_os`

File: robottelo/ui/discoveryrules.py [hunk 10]
```diff
@@ -28,17 +30,31 @@ def _configure_discovery(self, hostname=None, host_limit=None,
...
```
Comment (abalakh) [inline robottelo/ui/discoveryrules.py]: Can't see any usage of `select` arg. Is it redundant or implementation is missing?

File: robottelo/api/utils.py [hunk 11]
```diff
@@ -293,3 +303,246 @@ def create_role_permissions(role, permissions_types_names):  # pragma: no cover
...
+        org = entities.Organization().create()
+    if loc is None:
+        loc = entities.Location(organization=[org]).create()
+    # Create a new Life-Cycle environment
+    lc_env = entities.LifecycleEnvironment(organization=org).create()
+    # Create a Product, Repository for custom RHEL6 contents
+    product = entities.Product(organization=org).create()
+    repo = entities.Repository(
+        product=product,
+        url=settings.rhel7_os
```
Comment (sghai) [inline robottelo/api/utils.py]: @abalakh Raising an exception if that setting is commented out or none.

File: robottelo/config/settings.py (hunk 9 above)
Comment (sghai) [inline robottelo/config/settings.py]: updated.. thanks

File: robottelo/ui/discoveryrules.py (hunk 10 above)
Comment (sghai) [inline robottelo/ui/discoveryrules.py]: added implementation.. I just cherry-picked changes to fix tests.

**Pull request merged**